src/optimisation/Optimisation.jl

```diff
@@ -1,324 +1,327 @@
-        copyto!(G, ∇neglogp)
-
-        # If F is something, the negative log joint is requested as well.
-        # We have already computed it as a by-product above and hence return it directly.
-        if F !== nothing
-            return neglogp
-        end
-    end
+`lb`, `ub`, `cons`, `lcons`, and `ucons` are constraints for the optimization problem as
+described in `ModeEstimationConstraints` and Optimization.jl.
 
-    # Only negative log joint requested but no gradient.
-    if F !== nothing
-        return LogDensityProblems.logdensity(f, z)
-    end
+If `init_value` is `nothing`, it is generated by either sampling from the prior distribution
+or uniformly from the box constraints.
 
-    return nothing
+The problem is always created with `linked=false`. See [`link`](@ref) for how to transform
+it.
+"""
+function ModeEstimationProblem(
+    model, estimator, init_value, lb, ub, cons, lcons, ucons, adtype
+)
+    constraints = ModeEstimationConstraints(lb, ub, cons, lcons, ucons)
+    init_value = generate_init_value(model, init_value, constraints)
+    inner_context = estimator isa MAP ? DefaultContext() : LikelihoodContext()
+    ctx = OptimizationContext(inner_context)
+    log_density = OptimLogDensity(model, ctx)
+    return ModeEstimationProblem(
+        model, estimator, init_value, constraints, log_density, adtype, false
+    )
 end
 
+has_box_constraints(p::ModeEstimationProblem) = has_box_constraints(p.constraints)
+has_generic_constraints(p::ModeEstimationProblem) = has_generic_constraints(p.constraints)
 
+function default_solver(problem::ModeEstimationProblem)
+    return has_generic_constraints(problem.constraints) ? IPNewton() : LBFGS()
+end
 
-#################################################
-# Generic optimisation objective initialisation #
-#################################################
-
-function transform!!(f::OptimLogDensity)
-    ## Check link status of vi in OptimLogDensity
-    linked = DynamicPPL.istrans(f.varinfo)
+"""
+    link(p::ModeEstimationProblem)
 
-    ## transform into constrained or unconstrained space depending on current state of vi
-    f = Accessors.@set f.varinfo = if !linked
-        DynamicPPL.link!!(f.varinfo, f.model)
-    else
-        DynamicPPL.invlink!!(f.varinfo, f.model)
+Transform a mode estimation problem to unconstrained space, where all parameters can take
+any real value.
+"""
+function link(p::ModeEstimationProblem)
+    if has_constraints(p)
+        throw(
+            "Transforming constrained optimisation problems to unconstrained space is " *
+            "not yet implemented."
+        )
     end
-
-    return f
+    # Note that redefining obj and init_value out of place is intentional: It avoids
+    # issues with models for which linking changes the parameter space dimension.
+    ld = p.log_density
+    ld = Accessors.@set ld.varinfo = DynamicPPL.unflatten(ld.varinfo, copy(p.init_value))
+    ld = Accessors.@set ld.varinfo = DynamicPPL.link(ld.varinfo, ld.model)
```

Comment:
```julia
    ld = p.log_density
    varinfo = DynamicPPL.link(
        DynamicPPL.unflatten(ld.varinfo, copy(p.init_value))
        ld.model
    )
    ld = Accessors.@set ld.varinfo = varinfo
```

The constant `@set` just makes it unnecessarily difficult to read (know this is is from old code)

Reply:
Done